raco/backends/myria/myria.py

```diff
@@ -1216,6 +1301,22 @@ def __str__(self):
         return "Join => Shuffle(Join)"
 
 
+class ShuffleBeforeIDBController(rules.Rule):
+    def fire(self, expr):
+        # If not a join, who cares?
+        if not isinstance(expr, algebra.IDBController):
+            return expr
+        (group_list, agg) = expr.get_group_agg()
+
+        for idx in range(0, 2):
```

Comment:
The 3rd child is a consumer receiving EOS from `EOSController` so it's not a part of this rule (I should just use `range(2)` for sure). For shuffle vs. partial aggregation, it's not always correct to do partial aggregation before `IDBController`, since there are tricky aggs such as `CountFilter`. Another thing is it's not always more efficient to do partial aggregation before global. For example, if all the tuples are unique, then doing a paritial `DupElim` before an `IDBController` with `DupElim` would only double the cost. I think the difference here is that the aggs in `IDBController` and group-by are not the same ones.

I didn't check if eliding the shuffle is possible since I remember that redundant shuffles are removed by some other rules, but if it's not the case (or if it doesn't apply to `IDBController`) then I will add such a check here. (Is it?)